### PR TITLE
Add authn_method_add to v2 API

### PR DESCRIPTION
**Pull request by frederikrothenberger**
This PR adds the method `authn_method_add` to the II canister. This required to place some restrictions on the `metadata` entries that are mapped back to the `Device` type. However, once we migrate the stable memory to a type that more closely reflects the API v2, we should be able to lift this restriction.

### CODE REVIEW

File: src/internet_identity_interface/src/internet_identity/authn_method.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+use crate::internet_identity::types::{
+    AuthnMethod, AuthnMethodData, AuthnMethodProtection, PublicKeyAuthn, Purpose, WebAuthn,
+};
+use candid::Principal;
+use serde_bytes::ByteBuf;
+
+impl AuthnMethodData {
+    pub fn principal(&self) -> Principal {
```
Comment (przydatek) [inline src/internet_identity_interface/src/internet_identity/authn_method.rs:8]: Maybe add a comment describing the meaning of this `principal`?  It doesn't seem immediately obvious.

File: src/internet_identity_interface/src/internet_identity/authn_method.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        AuthnMethodData {
+            authn_method: AuthnMethod::PubKey(PublicKeyAuthn {
+                pubkey: ByteBuf::from(vec![0; 32]),
+            }),
+            metadata: Default::default(),
+            protection: AuthnMethodProtection::Unprotected,
+            purpose: Purpose::Authentication,
+            last_authentication: None,
+        }
+    }
```
Comment (przydatek) [inline src/internet_identity_interface/src/internet_identity/authn_method.rs:38]: These are very useful, but for testing -- should they be a part of the public API, or better added as separate test utils?

**Review by przydatek — APPROVED** [review]
LGTM, modulo some minor questions/suggestions. Thanks!

File: src/internet_identity_interface/src/internet_identity/authn_method.rs (hunk 2 above)
Comment (frederikrothenberger) [reply src/internet_identity_interface/src/internet_identity/authn_method.rs:38]: Yes, moved. 👍

File: src/internet_identity_interface/src/internet_identity/conversions.rs [hunk 3]
```diff
@@ -125,3 +137,89 @@ impl From<DeviceData> for AuthnMethodData {
...
+                            key: key.to_string(),
+                            expected_type: "string".to_string(),
+                            actual_type: "bytes".to_string(),
+                        })
+                    }
+                    MetadataEntry::Map(_) => Err(AuthnMethodConversionError::InvalidMetadataType {
+                        key: key.to_string(),
+                        expected_type: "string".to_string(),
+                        actual_type: "map".to_string(),
+                    }),
```
[the remark this answers is not in the document]
Comment (przydatek) [reply src/internet_identity_interface/src/internet_identity/conversions.rs:197]: Ah, yes, you are right, sorry for the noise.

[the remark this answers is not in the document]
Comment (frederikrothenberger) [reply src/internet_identity_interface/src/internet_identity/conversions.rs]: TIL about `@` bindings. Thanks.

[the remark this answers is not in the document]
Comment (przydatek) [reply src/internet_identity_interface/src/internet_identity/conversions.rs]: Oh, me too! Thank you! 😄

**Pull request merged**